matching: ellipsis: Always try going deep

IagoAbal opened the pull request:
Pattern `... bar()` would behave strangely because it would only try to match deeply if it did not match anything non-deeply. So depending on how the target code looked, `... bar()` would match or not match inside e.g. an `if` statement. This was confusing and not a great property for Semgrep to have.

The regression was introduced in PR #852 as part of a set of optimizations done back in 0.9.0 (!), but at present reverting this one thing does not seem to have any negative perf impact.

Follows: c1ca429c183 ("Optimize deep statement matching (#852)")

Closes returntocorp/semgrep-rules#660
Closes PA-2992

test plan:
make test # new tests
Also, compared against develop running p/default on 32 repos from stress-test-monorepo, and no meaningful slowdown or increase in memory usage was observed.

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

IagoAbal commented:
The CI benchmarks failing is unrelated, they fail on every PR.

I'll open a PR to semgrep-rules to fix some test expectations, but first I want to get a review on this PR to be sure it will be merged.

aryx commented on src/matching/Generic_vs_generic.ml:
```diff
@@ -2237,15 +2237,14 @@ and m_stmts_deep ~inside ~less_is_ok (xsa : G.stmt list) (xsb : G.stmt list) =
   | ( ({ s = G.ExprStmt ({ e = G.Ellipsis _i; _ }, _); _ } :: _ as xsa),
       (_ :: _ as xsb) ) ->
-      (* let's first try without going deep *)
-      m_list__m_stmt xsa xsb >!> fun () ->
```
each time we use >!>, there is a bug because we cut some matches.
Ideally we should remove this combinator.
Ideally we should use CTL instead of those deep stmt hacks.

IagoAbal commented:
Rebased and don't know what happened... will look into it.

IagoAbal committed:
matching: ellipsis: Always try going deep

Follows: c1ca429c183 ("Optimize deep statement matching (#852)")

test plan:
make test # new tests
